File: node/mods/rustscript/lib/opcodes/setfield.js
/**
 * Purpose: SETFIELD opcode — copy a value into a writable context location.
 *
 * Language shape:
 *   script.reference  — destination path (must be context.*)
 *   script.value      — literal or resolve_ref source
 *
 * Read-only destinations (script.*, witness.*, tx.*, blk.*) are rejected.
 */

module.exports = {
  name: 'SETFIELD',
  description: `
Copies a value into a writable location in the execution context.

reference is the destination path. value is a literal or a normal RustScript
reference resolved like every other opcode operand.

Only context.* destinations are writable.
`,
  exampleScript: {
    op: 'SETFIELD',
    reference: 'context.constitution.owner',
    value: '__opcodes.importfield.owner'
  },
  schema: {
    reference: 'string',
    value: 'reference'
  },

  execute(node, context) {
    if (!node || typeof node !== 'object' || !context || typeof context !== 'object') {
      return false;
    }

    const reference = node.reference;
    if (typeof reference !== 'string' || reference.length === 0) {
      return false;
    }

    if (!Object.prototype.hasOwnProperty.call(node, 'value')) {
      return false;
    }

    if (!reference.startsWith('context.')) {
      return false;
    }

    const path = reference.slice('context.'.length);
    if (!path || isForbiddenWritePath(path)) {
      return false;
    }

    const value = resolveRef(context, node.value);
    return setContextPath(context, path, value);
  }
};

function isForbiddenWritePath(path) {
  return (
    path === 'script' ||
    path.startsWith('script.') ||
    path === 'witness' ||
    path.startsWith('witness.') ||
    path === 'tx' ||
    path.startsWith('tx.') ||
    path === 'blk' ||
    path.startsWith('blk.')
  );
}

function pathSegments(path) {
  return String(path)
    .replace(/\[/g, '.')
    .replace(/\]/g, '')
    .split('.')
    .filter((s) => s.length > 0);
}

function setContextPath(root, path, value) {
  const segments = pathSegments(path);
  if (!segments.length) {
    return false;
  }

  let current = root;
  for (let i = 0; i < segments.length; i += 1) {
    const segment = segments[i];
    const isLast = i === segments.length - 1;
    const asIndex = /^[0-9]+$/.test(segment) ? Number(segment) : null;

    if (isLast) {
      if (asIndex !== null) {
        if (!Array.isArray(current) || asIndex >= current.length) {
          return false;
        }
        current[asIndex] = value;
        return true;
      }
      if (!current || typeof current !== 'object' || Array.isArray(current)) {
        return false;
      }
      current[segment] = value;
      return true;
    }

    if (asIndex !== null) {
      if (!Array.isArray(current) || asIndex >= current.length) {
        return false;
      }
      current = current[asIndex];
      continue;
    }

    if (!current || typeof current !== 'object' || Array.isArray(current)) {
      return false;
    }
    if (!Object.prototype.hasOwnProperty.call(current, segment)) {
      current[segment] = {};
    }
    current = current[segment];
  }

  return false;
}

function resolveRef(root, ref) {
  if (typeof ref !== 'string') {
    return ref;
  }
  const parts = ref.split('.');
  let cursor = root;
  for (let i = 0; i < parts.length; i += 1) {
    const key = parts[i];
    if (
      !cursor ||
      typeof cursor !== 'object' ||
      !Object.prototype.hasOwnProperty.call(cursor, key)
    ) {
      return ref;
    }
    cursor = cursor[key];
  }
  return cursor;
}

File: rust/saito-core/src/core/consensus/scripting/opcodes/setfield.rs
use crate::core::consensus::block::Block;
use crate::core::consensus::transaction::Transaction;
use serde_json::Value;

use super::super::script::resolve_ref;

/// SETFIELD — copy a resolved value into a writable location in the
/// execution context.
///
/// Script shape:
/// ```json
/// {
///   "op": "SETFIELD",
///   "reference": "context.constitution.owner",
///   "value": "__opcodes.importfield.owner"
/// }
/// ```
///
/// Destinations must be `context.*` paths. Writes to `script.*`, `witness.*`,
/// `tx.*`, and `blk.*` are rejected.
pub struct SetField {
    pub name: String,
    pub description: String,
    pub script: String,
    pub schema: Value,
}

fn is_forbidden_write_path(path: &str) -> bool {
    path == "script"
        || path.starts_with("script.")
        || path == "witness"
        || path.starts_with("witness.")
        || path == "tx"
        || path.starts_with("tx.")
        || path == "blk"
        || path.starts_with("blk.")
}

/// Normalize `foo[0].bar` into dotted segments compatible with resolve_ref lookup.
fn path_segments(path: &str) -> Vec<String> {
    let normalized = path.replace('[', ".").replace(']', "");
    normalized
        .split('.')
        .filter(|s| !s.is_empty())
        .map(|s| s.to_string())
        .collect()
}

fn set_context_path(root: &mut Value, path: &str, value: Value) -> bool {
    let segments = path_segments(path);
    if segments.is_empty() {
        return false;
    }

    let last = segments.len() - 1;
    let mut current = root;

    for (i, segment) in segments.iter().enumerate() {
        if i == last {
            if let Ok(index) = segment.parse::<usize>() {
                let Some(arr) = current.as_array_mut() else {
                    return false;
                };
                if index >= arr.len() {
                    return false;
                }
                arr[index] = value;
                return true;
            }

            if !current.is_object() {
                *current = Value::Object(serde_json::Map::new());
            }
            if let Value::Object(map) = current {
                map.insert(segment.clone(), value);
                return true;
            }
            return false;
        }

        if let Ok(index) = segment.parse::<usize>() {
            let Some(arr) = current.as_array_mut() else {
                return false;
            };
            if index >= arr.len() {
                return false;
            }
            current = &mut arr[index];
            continue;
        }

        if !current.is_object() {
            *current = Value::Object(serde_json::Map::new());
        }

        let Value::Object(map) = current else {
            return false;
        };

        if !map.contains_key(segment) {
            map.insert(segment.clone(), Value::Object(serde_json::Map::new()));
        }

        current = map.get_mut(segment).unwrap();
    }

    false
}

impl SetField {
    pub fn validate(context: &mut Value, tx: Option<&Transaction>, blk: Option<&Block>) -> u8 {
        let reference = context["script"]["reference"]
            .as_str()
            .unwrap_or("")
            .to_string();
        if reference.is_empty() {
            return 0;
        }

        let Some(path) = reference.strip_prefix("context.") else {
            return 0;
        };
        let path = path.to_string();
        if path.is_empty() || is_forbidden_write_path(&path) {
            return 0;
        }

        if context["script"].get("value").is_none() {
            return 0;
        }

        let value = resolve_ref(&context["script"]["value"].clone(), context, tx, blk);

        if set_context_path(context, &path, value) {
            1
        } else {
            0
        }
    }
}
